### ISSUE TITLE
FN: needless_collect: collecting without turbofish 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6932]
Lint name: needless_collect

I tried this code:

```rust
struct A {
    i: i32,
}

fn fun(numbers: &[i32]) -> impl Iterator<Item = A> {
    let v: Vec<_> = numbers.iter().map(|x| A { i: *x }).collect();
    v.into_iter()
}
```

does not trigger the lint while
````rust

struct A {
    i: i32,
}

pub fn fun(numbers: &[i32]) -> impl Iterator<Item = A> {
    let v = numbers.iter().map(|x| A { i: *x }).collect::<Vec<A>>();
    v.into_iter()
}
````
does

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (36f1f04 2021-03-17)`
- `rustc -Vv`:
  ```
  rustc 1.46.0-nightly (f455e46ea 2020-06-20)
  binary: rustc
  commit-hash: f455e46eae1a227d735091091144601b467e1565
  commit-date: 2020-06-20
  host: x86_64-unknown-linux-gnu
  release: 1.46.0-nightly
  LLVM version: 10.0

  ```

**Comment by camsteffen** [comment]
Fixed by #7163
